content/blog/2025/07-01_httr2_client/index.qmd

```diff
@@ -0,0 +1,623 @@
+    ),
+    {
+      expect_false(ns_has_auth())
+    }
+  )
+})
+```
+
+Here I am not mocking, but explicitly setting environment variables for the tests with `withr::with_envvar`.
+This way, the code is tested as I want, regardless of whether system running the tests has the variables set (as I have on my system).
+
+## Periodically testing API Calls
+
+In the tests so far, I've recorded the API requests and responses on my local machine where credentials are set up, and then all subsequent tests run against the recorded cassettes.
+But how do we discover if the API it self alters?
+Since the API calls aren't actually executed, I'd need to rerun the tests without the cassettes
+But we can be sure I won't do that unless someone reports errors, and ideally I'd like to be a little more pro-active than that.
+
+Thankfully, vcr has a way to let us do that!
+If you set an environment variable `VCR_TURN_OFF=true` in your `.Renviron`, then the tests will run where the cassettes are not used.
+Neat!
+Of course they would provide such a nice and neat solution to do this.
+So I made a GitHub Action workflow that will run weekly and test against the API, so I make sure that things work.
+I will only be notified on e-mail if it fails, so I can just forget about it until I get a notification.
+Which is pretty nice, in my opinion.
+
+```yaml
+# Workflow derived from https://github.com/r-lib/actions/tree/v2/examples
+# Need help debugging build failures? Start at https://github.com/r-lib/actions#where-to-find-help
+on:
+  push:
+    branches: [main]
+  pull_request:
+  schedule:
+    - cron: "30 2 * * 0"
+
+name: Check API endpoints
+
+permissions: read-all
+
+jobs:
+  R-CMD-check:
+    runs-on: ubuntu-latest
+
+    name: Check API responses
+
+    env:
+      GITHUB_PAT: ${{ secrets.GITHUB_TOKEN }}
+      R_KEEP_PKG_SOURCE: yes
+
+    steps:
+      - uses: actions/checkout@v4
+
+      - uses: r-lib/actions/setup-pandoc@v2
+
+      - uses: r-lib/actions/setup-r@v2
+        with:
+          r-version: "release"
+          http-user-agent: "release"
+          use-public-rspm: true
+
+      - uses: r-lib/actions/setup-r-dependencies@v2
+        with:
+          extra-packages: |
+            any::rcmdcheck
+            any::knitr
+          needs: check
+
+      - name: "Setup env for full test suite"
+        run: |
+          echo "NETTSKJEMA_CLIENT_ID=${{ secrets.CLIENT_ID }}" >> .Renviron
+          echo "NETTSKJEMA_CLIENT_SECRET=${{ secrets.CLIENT_SECRET }}" >> .Renviron
+          echo "VCR_TURN_OFF=true" >> .Renviron
+
+      - uses: r-lib/actions/check-r-package@v2
+        with:
+          upload-snapshots: true
+          build_args: 'c("--no-manual","--compact-vignettes=gs+qpdf","--as-cran")'
+
+      - name: Build vignettes
+        run: |
+          install.packages(".", repos = NULL, type = "source")
+          nettskjemar:::knit_vignettes()
+        shell: Rscript {0}
+```
+
+## Making vignettes that call the API
```

Comment:
https://docs.ropensci.org/vcr/articles/vcr.html#other-uses-examples-and-vignettes